docs/typing-indicators.md

```diff
@@ -0,0 +1,146 @@
+If the "writing user" is composing a long message, we want to send
+repeated updates to the server, so that downstream clients know that the
+user is still typing.  (Zulip messages tend to be longer than
+messages in other chat/text clients.)
+
+We have a small state machine in `static/js/typing_status.js` that
+makes sure subsequent "start" requests get sent out every ten
+seconds.  (This document is intended to describe the high level
+architecture; the actual time values may be tuned in future releases.)
+
+If the "writing user" goes more than five seconds without any text
+input, then we send a request with an `op` of `stop`.  We also send
+"stop" messages when the user closes the compose box.
+
+A common scenario is that a user is just pausing to think for a few
+seconds, but they still intend to finish the message.  Of course,
+that's hard to distinguish from the scenario of the user got pulled
+away from their desk.  For the former case, where the "writing user"
+completes the message with lots of pauses for thinking, a series of
+"start" and "stop" messages may be sent over time.  One way to
+improve our feature here is to tweak the timeout values, but of
+course it is always a tradeoff depending on how quickly people type,
+how long they pause to think, how frequently they get interrupted, etc.
+
+## Server
+
+The server piece of typing notificiations is currently pretty
+straightforward, since we take advantage of Zulip's event system,
+and the server piece here is basically a single Django view
+function with a small bit of library support to send out events
+to clients.
+
+Requests come into `/json/typing`.  The view mostly calls out
+to `check_send_typing_notification` to do the heavy lifting.
+
+One of the main things that the server does is to simply validate
+the recipients with a call to `recipient_for_emails`.  (We should
+streamline the payload in the request to have user ids instead of
+emails, but of course we will still need to validate recipients.)
+
+Once the request has been validated, the server sends events to
+potential recipients of the message.  The event type for that
+payload is `typing`.  See the function `do_send_typing_notification`
+for more details.
+
+## Reading user
+
+When a user plays the role of a "reading user," the client handles
+incoming "typing" events from the server, and the client will
+display typing notification only when both of these conditions are
+true:
+
+* The "writing user" is still likely typing.
+* The "reading user" is in a view where they'd see the eventual
+  message.
+
+The simplest case to consider is that the "reading user" is narrowed
+to a 1:1 conversation with the "writing user."  As the typing
+notification events come in to the client, a status message is shown
+just below the message feed.  It's possible for the "reading user"
+to be in a broader PM view, too, and when that happens, we look
+at the incoming events for all potential users that can be typing
+messages to us.
```

Comment:
I would probably explain this a bit differently; that the "receiving user" is responsible for maintaining the data as to which views currently have which users typing based on the notices coming from other clients, which we do via `typing_events.js`, since I think that's what is important to explain for someone working on the mobile implementation, for example.  

And then maybe explain how we display that data in the view code a bit further down.